05_testing_and_ci/github_actions_gitlab_ci_demo.md

```diff
@@ -0,0 +1,412 @@
+- Checkout `Dockerfile` in `examples/automation` directory
+    - Image is build on top of [Alpine Linux](https://www.alpinelinux.org). Alpine is optimized to produce tiny build to be used in containers.
+    - We fix the version of the base image to make sure to not run into compatibility problems.
+    - We need to install Python etc. ourselves:
+        - `python3`, `pip` and Python libraries/packages (`black`)
+        - `pip` should be installed and used from a non-root user. Therefore, we create a user called `testuser` and change to this user
+    - Password-less sudo is enabled for the user in case it would be needed to install other packages.
+
+- Build the image locally and tag it according to my DockerHub account name (`ajaust`)
+
+  ```bash
+  sudo docker build -t ajaust/automation-lecture .
+  ```
+
+- Checkout the image list `sudo docker image list` to point out that the container is reasonably small. During the preparation it was about 72 MB large/small. This about the size of the size of the pure Ubuntu base container:
+
+  ```text
+  ubuntu                          20.04        ba6acccedd29   3 months ago   72.8MB
+  ajaust/automation-lecture       latest       a0dec678c941   5 hours ago    71.2MB
+  ```
+
+- Push image to Docker Hub such that we can use it in CI pipelines
+
+  ```bash
+  sudo docker push ajaust/automation-lecture
+  ```
+
+- Now the image can be used as `ajaust/automation-lecture` or `ajaust/automation-lecture:latest`. I did not version the uploaded container.
+
+## GitHub Actions
+
+### 1. Setting up a test job
+
+- Go to [repository](https://github.com/Simulation-Software-Engineering/automation-lecture) an check out tag `initial-state`.
+- Create a new branch `lecture-branch`
+- Set up workflow file
+
+  ```bash
+  mkdir -p .github/workflows
+  cd .github/workflows
+  touch testing.yml
+  ```
+
+- Run `black` shortly locally and explain what it does.
+
+  ```bash
+  black --check .
+  ```
+
+    - It is a style checker that checks whether the Python files are formatted according to a certain standard.
+    - `black --check .` will check files without modifying them.
+    - `black` uses a lot of smileys
+
+- Edit `testing.yml` to have following content
+
+  ```yaml
+  name: Testing workflow
+
+  on: [push, pull_request]
+
+  jobs:
+    test:
+      runs-on: ubuntu-latest
+      steps:
+        - uses: actions/checkout@v2
+        - uses: actions/setup-python@v2
+          with:
+            python-version: '3.8.10'
+        - name: "Install pytest dependency"
+          run: pip install --user  pytest
+        - name: "Run unittest"
+          run: python -m unittest
+  ```
+
+- `runs-on` does **not** refer to a Docker container, but to runner tag.
+- After the push inspect "Action" panel on GitHub repository
+    - GitHub will schedule a run (yellow dot)
+    - Hooray. We have set up our first action.
+- Edit `test_operations.py`, break a test, commit it and push it to the repository. Check the failing test.
```

Comment:
Actually, it could be nicer from a reality viewpoint to break the implementation, not the test. Add to do it in a PR? And maybe to set the options in the repo that you cannot merge then?